### ISSUE TITLE
[预训练模型] 钙钛矿材料性质预测模型

### ISSUE THREAD

**Issue by newtontech** [newtontech/Perovskite_Pretrain_Models#5]
## 研究背景

机器学习可以加速材料性质预测，减少实验次数。

## 问题描述

当前预训练模型需要扩展：
1. 预测更多性质
2. 提高预测精度
3. 支持新组分

### 1. 多任务学习框架

```python
# models/perovskite_property_predictor.py
import torch
import torch.nn as nn

class PerovskitePropertyPredictor(nn.Module):
    def __init__(self, input_dim, hidden_dims):
        super().__init__()
        
        # 共享编码器
        self.encoder = nn.Sequential(
            nn.Linear(input_dim, hidden_dims[0]),
            nn.ReLU(),
            nn.Linear(hidden_dims[0], hidden_dims[1]),
            nn.ReLU()
        )
        
        # 多任务头
        self.pce_head = nn.Linear(hidden_dims[1], 1)
        self.bandgap_head = nn.Linear(hidden_dims[1], 1)
        self.stability_head = nn.Linear(hidden_dims[1], 1)
    
    def forward(self, composition_features):
        encoded = self.encoder(composition_features)
        return {
            'pce': self.pce_head(encoded),
            'bandgap': self.bandgap_head(encoded),
            'stability': self.stability_head(encoded)
        }
```

### 2. 图神经网络模型

```python
# models/perovskite_gnn.py
import torch_geometric
from torch_geometric.nn import GCNConv, global_mean_pool

class PerovskiteGNN(nn.Module):
    """用于晶体结构预测的图神经网络"""
    def __init__(self, node_dim, edge_dim, hidden_dim):
        super().__init__()
        self.conv1 = GCNConv(node_dim, hidden_dim)
        self.conv2 = GCNConv(hidden_dim, hidden_dim)
        self.pool = global_mean_pool
        self.predictor = nn.Linear(hidden_dim, 1)
    
    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = F.relu(self.conv1(x, edge_index))
        x = F.relu(self.conv2(x, edge_index))
        x = self.pool(x, data.batch)
        return self.predictor(x)
```

### 3. 预测任务列表

| 性质 | 输入 | 输出 | 用途 |
|------|------|------|------|
| 带隙 | 组分 | eV | 叠层设计 |
| PCE | 组分+工艺 | % | 效率预测 |
| T80 | 组分+条件 | h | 稳定性 |
| 形成能 | 组分 | eV | 合成可行性 |
| 缺陷密度 | 组分+工艺 | cm⁻³ | 质量预测 |

## 验收标准

- [ ] 实现 PCE 预测（MAE <2%）
- [ ] 实现带隙预测（MAE <0.1 eV）
- [ ] 实现稳定性预测
- [ ] 模型可解释性分析

## 预期成果

- 预训练材料预测模型
- 在线预测 API
- 特征重要性分析

## 相关文件

- models/
- data/training/
- notebooks/

**Comment by newtontech** [comment]
Progress update: PR #7 was merged with a deterministic property-prediction scaffold, CSV target loading, sparse-target handling, kNN prediction, feature-importance output, and focused tests. I am keeping this issue open because the publication-grade acceptance targets still need real data/training: PCE MAE <2%, bandgap MAE <0.1 eV, stability quality, and interpretability on real benchmarks.

**Comment by newtontech** [comment]
Additional progress update: PR #9 has been merged, adding a complete, runnable property prediction demo with multiple property support, feature importance analysis, and benchmark evaluation framework.

**What's now implemented:**
✅ Executable property prediction workflow (property_prediction_demo.py)
✅ Support for multiple properties: PCE, bandgap, stability (T80)
✅ Feature importance analysis and interpretability
✅ Benchmark evaluation framework with leave-one-out cross-validation
✅ Molecular generation and screening workflow
✅ Comprehensive documentation and examples

**What remains for full issue completion:**
❌ Real perovskite dataset integration (currently using synthetic data)
❌ Full model training with pre-trained models (Uni-Mol/MolCLR)
❌ Validation of MAE targets: PCE MAE <2%, bandgap MAE <0.1 eV
❌ Online prediction API (as mentioned in original issue)

The current implementation provides a solid, reproducible framework that can be used with real datasets once they are available. The demo uses deterministic models and synthetic data to ensure the code is testable and verifiable.

To complete this issue, the following steps are needed:
1. Integrate real perovskite property datasets
2. Train full models using the existing train/run.py and train/train_molclr/finetune.py pipelines
3. Validate performance meets the MAE targets on real data
4. Deploy online prediction API if needed

The framework is ready for production use with real datasets.

**Comment by newtontech** [comment]
## Issue #5 Status: Infrastructure Complete, Training Metrics Pending

**Assessment date:** 2026-06-08
**Reviewer:** Fleet worker 5.1

---

### Acceptance Criteria Analysis

| Criterion | Status | Evidence |
|-----------|--------|----------|
| **PCE prediction (MAE <2%)** | Infrastructure ready, threshold not validated on real data | `perovskite_pretrain/property_prediction.py` implements multi-target prediction with MAE computation. Deterministic kNN scaffold runs on synthetic data (demo MAE ~0.36 on synthetic PCE). No trained model meets the <2% MAE target on real perovskite data. |
| **Bandgap prediction (MAE <0.1 eV)** | Schema stub, threshold not validated | Config `configs/pretraining/property_prediction_multitask.json` defines bandgap target with `status: schema_stub`. Column `BANDGAP_EV` is wired but no trained model achieves <0.1 eV MAE. |
| **Stability prediction** | Schema stub | `stability_t80` target defined in config with `acceptance_target: dataset_defined`. T80 column wired through `T80_HOURS`. No trained model or real benchmark yet. |
| **Model interpretability** | Implemented for tabular features | `DeterministicPropertyPredictor.feature_importance()` computes per-task feature-target covariance. Demo shows ranked feature output. Attention heatmap and embedding UMAP are documented in the config under `evaluation.interpretability` but require trained model artifacts. |

### What IS Implemented and Tested

1. **Property prediction infrastructure** — `perovskite_pretrain/property_prediction.py`:
   - CSV loading with sparse multi-target support (`load_property_rows`)
   - Deterministic kNN regressor (`DeterministicPropertyPredictor`)
   - Regression metrics: MAE, RMSE, R2 (`regression_metrics`)
   - Leave-one-out benchmark with acceptance thresholds (`leave_one_out_benchmark`)
   - JSON report generation with dataset manifest (`write_benchmark_report`)

2. **Multi-task prediction** — Three targets (delta_pce, bandgap, stability_t80) defined in config and code. Each target independently loadable, predictable, and benchmarked.

3. **GNN support** — Existing codebase contains full MolCLR GNN fine-tuning pipeline at `train/train_molclr/` (GCN and GIN models) and Uni-Mol transformer at `train/unimol_tools/`. The property prediction config references pretrained molecular encoders.

4. **Interpretability** — Feature importance analysis implemented. Config specifies attention heatmaps and embedding UMAP as planned interpretability outputs.

5. **Executable demo** — `examples/notebooks/property_prediction_demo.py` runs end-to-end with `python3 examples/notebooks/property_prediction_demo.py` (verified locally).

6. **Tests** — 9/9 tests pass locally (`tests/test_deterministic_skeletons.py` + `tests/test_static_smoke.py`).

7. **CI** — Latest main CI run is green (commit `feat: add executable property prediction demo and examples`).

8. **Documentation** — `docs/pretraining_workflows.md` covers all three issues (#3, #5, #6) with reproducible workflow instructions.

### What Remains Blocked

The MAE acceptance thresholds (PCE <2%, bandgap <0.1 eV) require:
- Real curated perovskite property datasets (not synthetic demo data)
- Full model training runs using Uni-Mol / MolCLR / ChemBERTa2 pipelines
- Benchmark publication with dataset manifest, split seeds, and artifact checksums

These are data-dependent and compute-intensive steps that cannot be honestly completed without real datasets and training cycles.

### Recommendation

**Do NOT close this issue.** The code infrastructure is complete and well-tested, but the scientific acceptance criteria require real data and trained model benchmarks. Suggested follow-up issues:

1. **Dataset manifest** — Document and integrate real perovskite property data with PCE, bandgap, and T80 labels
2. **Model training run** — Execute full training pipeline with Uni-Mol / MolCLR on real data across all 5 split seeds
3. **MAE benchmark publication** — Publish calibrated MAE results with dataset manifest, config, and artifact checksums

Keeping this issue open until real benchmarks validate the MAE thresholds.
